sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1244,6 +1244,23 @@ class Analyzer(
         case s @ ScalarSubquery(sub, conditions, exprId)
             if sub.resolved && conditions.isEmpty && sub.output.size != 1 =>
           failAnalysis(s"Scalar subquery must return only one column, but got ${sub.output.size}")
+        // SPARK-1854: block cases where GROUP BY columns
+        // are not part of the correlated columns
+        case s @ ScalarSubquery(sub, conditions, exprId)
```

Comment:
We also do a part of the scalar subquery checks in CheckAnalysis. Lets move this code there.